plugins/kubernetes/app/models/kubernetes/build_job_executor.rb

```diff
@@ -0,0 +1,140 @@
+# Execute a job that can build/push Docker images
+# and write progress to the local job output.
+module Kubernetes
+  class BuildJobExecutor
+    WAIT_FOR_JOB = 5.minutes
+    TICK = 3.seconds
+
+    attr_reader :job
+    delegate :id, to: :job
+
+    def initialize(output, job:)
+      @output = output
+      @job = job
+      @registry = {}
+    end
+
+    def execute!(build, project, tag:, push: false, registry:, tag_as_latest: false)
+      job_log = job_name = job_namespace = ""
+      k8s_job = nil
+
+      unless valid_registry_info?(registry)
+        @output.puts "### Registry server should not be empty. Aborting..."
+        return false, job_log
+      end
+      @registry = registry
+
+      # Read the external config path and create a new job config instance
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::RoleConfigFile.new(contents, build_job_config_path).job
+      fill_job_details(k8s_job, build, project, tag: tag, push: push, tag_as_latest: tag_as_latest)
+
+      if k8s_job
+        job_name = k8s_job[:metadata][:name]
+        job_namespace = k8s_job[:metadata][:namespace]
+      end
+
+      @output.puts "### Running a remote Docker build job: #{job_name}"
+      success, job_log = create_and_wait_for_job(k8s_job, @output)
+
+      if success
+        @output.puts "### Remote build job #{job_name} completed sucessfully"
+      else
+        @output.puts "### Remote build job #{job_name} failed to run or timed out"
+      end
+
+      return success, job_log
+    ensure
+      if k8s_job
+        # Jobs will still be there regardless of their statuses
+        @output.puts "### Cleaning up the remote build job: #{job_name}"
+        extension_client.delete_job(job_name, job_namespace)
```

Comment:
Yes, if it fails to create it will just throw 404 and be caught by the outer try/catch block of job executor, sending a proper message to the log. This ```ensure``` block is the best effort attempt to clean up resource remnants from the build job, what else do you think is good to add here?